rfc: LDN

liz3 opened the pull request:
# What is this
This came to me out of the idea and complains i heard from a few friends saying their systems wouldnt handle Lost ark as a game and loa details, now i dont think the code of this project is the issue but it is true its spawns two electron windows, which means running 2 chromium instances extra.  
LDN = Loa details native, idk can also suggest name changes

This is a very simplistic "alternative" damage meter window, targeted to directly render on the GPU, it has a very basic ui system for rendering fonts and rectangles and uses GLFW to target OpenGL.

# Motivation
Part of the motivation is already explained above, but the other reason is just that it seamed fun to build.

# Progress
**THIS IS WIP** Theres no way around that, that this isnt ready to be released, im opening the comment as the title says as a request for comments and feedback.

Supported tabs: (All tabs included by default in 
* Damage tab(no skill breakdown atm)
* rdps tab
* tank tab
* Party synergies tab
* Shield Given tab(SHIELD D)
* Effective shield given(E SHIELD D)
* Self buff damage(SBDMG)

Supported features:
* Taking screenshot of the meter(without names)
* passthrough for mouse events(ghost mode)
* minimising
* pausing the meter

 It sits between 30-50mb ram and 0 cpu usage unless you interact with the windows.

It is is designed to look very similar to the main damage overlay window.

# How does it work
I added a extra icon button to the main damage meter window which when pressed closes both electron windows(hides main rn because otherwise quasar would quit) and then spaws a localhost tcp socket server and starts the executable with that port who then connects to that server and starts receiving the live parsers state. the logic used to determine dps and buffs is sadly duplicated and reimplemented in c++, since the target is to entirely disable the two electron windows leaving only the node process. 

# How to run this
1. You will need CMake and a installation of visual studios c++ build chain supporting c++ 17.
2. Check out the submodules.
3. in the root of this branch go to the `ldn` folder, there create a folder called `build_release`
4. Open a terminal in that folder and run `cmake .. -DCMAKE_BUILD_TYPE=Release` after that `cmake --build . --config Release`
5. From there you *should* be able to run dev or build and use the button, emphasis on should because im sure theres stuff which can go wrong. If it works, rn just close the window in the window bar and it will restart the electron windows. 

## Some other stuff

* Font is weird: currently its a hardcoded size and im working on pretty big screens, will adjust
* Something does not work at all: leave a comment



![image](https://github.com/lost-ark-dev/loa-details/assets/21298625/b5d566cf-ad77-47a3-866b-266d2e48f112)
![image](https://github.com/lost-ark-dev/loa-details/assets/21298625/ae9d6ce9-1f79-4d08-acb1-ac4c26f4e460)
![image](https://github.com/lost-ark-dev/loa-details/assets/21298625/6f351f2c-d119-4d9c-9cc3-dabfbac2efc4)
![image](https://github.com/lost-ark-dev/loa-details/assets/21298625/6c64a215-6755-45df-8f32-c88a95b49aa4)
![image](https://github.com/lost-ark-dev/loa-details/assets/21298625/a5dd5d1e-4f93-43ad-8016-0a99653c3cd5)
![image](https://github.com/lost-ark-dev/loa-details/assets/21298625/fd5dcb88-6b03-4049-9fad-84322bf26bfb)

liz3 committed:
feat: ldn

Herysia commented:
Tbh i think that it'd more appropriate to write a "network dispatch" for meter-core and run meter core headless with node. So that any app could use that network dispatch api to display it the way they like.

Loa details is the web based eletron rendered ui for the dps meter. I don't think that there is any reason to add a native replacement inside this repo.

Regarding loa details performance, i think that more than electron itself the ui code is garbage. Skill issue i guess 🤣...

liz3 committed:
fix: remove deps as source code and add them as submodules

liz3 committed:
chore: also include json lib as submodule

liz3 committed:
fix: correctly empty data_buffer incase of partial messages

liz3 commented:
> Regarding loa details performance, i think that more than electron itself the ui code is garbage. Skill issue i guess 🤣...

I do not think thats true, of course source code and ui code can always be optimised but running a browser just has a upfront overhead due to its nature, non of the opengl code is really optimised in this either currently and it can just be lower since its the code only and no overhead otherwise.

> Loa details is the web based eletron rendered ui for the dps meter. I don't think that there is any reason to add a native replacement inside this repo.

As outlined its not meant as a replacement to replace the main electron one but was intended as a second option for users with lower specs/generally a faster window, building ui for this is just a lot harder then using vue.

> Tbh i think that it'd more appropriate to write a "network dispatch" for meter-core and run meter core headless with node. So that any app could use that network dispatch api to display it the way they like.

Thats something i can look into it sounds like a good idea yeah

liz3 committed:
fix: unhardcode images path and load classes from databases

liz3 committed:
fix: disable linter here

liz3 committed:
feat: add tank and rdps tabs

liz3 committed:
feat: window mouse events passthrough

liz3 committed:
feat: add shield g, e sheld g and sdbuff tabs

Also added a webp decoder because it turns out that there are webps in meter-data.

liz3 committed:
feat: hide names when taking screenshots

liz3 committed:
feat: pausing/reset session

liz3 committed:
fix: change some json instances to be refs and actually close main window

liz3 committed:
fix: remove this annoying include

liz3 committed:
chore: rename tab name for party synergy

liz3 committed:
Merge remote-tracking branch 'origin/main' into liz3/ldn

liz3 committed:
feat: icons, proper scaling for fonts now and some more stuff

liz3 committed:
fix: crash when setname is not present even though categorytype is set

liz3 committed:
fix: empty boss name if its not present

liz3 committed:
feat: add gear score to name display
